marlowe-playground-client/src/MarloweEditor/State.purs

```diff
@@ -0,0 +1,222 @@
+  liftEffect $ LocalStorage.setItem marloweBufferLocalStorageKey contents
+  assign _hasUnsavedChanges' false
+
+handleAction _ MarkProjectAsSaved = assign _hasUnsavedChanges' false
+
+handleAction _ (SelectHole hole) = assign _selectedHole hole
+
+handleAction settings AnalyseContract =
+  void
+    $ runMaybeT do
+        contents <- MaybeT $ editorGetValue
+        contract <- hoistMaybe $ parseContract' contents
+        -- FIXME: when editor and simulator were together the analyse contract could be made
+        --        at any step of the simulator. Now that they are separate, it can only be done
+        --        with initial state
+        let
+          emptySemanticState = emptyState zero
+        assign _analysisState (WarningAnalysis Loading)
+        response <- lift $ checkContractForWarnings contract emptySemanticState
+        assign _analysisState (WarningAnalysis response)
+  where
+  checkContractForWarnings contract state = runAjax $ (flip runReaderT) settings (Server.postMarloweanalysis (MSReq.Request { onlyAssertions: false, contract, state }))
+
+handleAction settings AnalyseReachabilityContract =
+  void
+    $ runMaybeT do
+        contents <- MaybeT $ editorGetValue
+        contract <- hoistMaybe $ parseContract' contents
+        -- FIXME: when editor and simulator were together the analyse contract could be made
+        --        at any step of the simulator. Now that they are separate, it can only be done
+        --        with initial state
+        let
+          emptySemanticState = emptyState zero
+        newReachabilityAnalysisState <- lift $ startReachabilityAnalysis settings contract emptySemanticState
+        assign _analysisState (ReachabilityAnalysis newReachabilityAnalysisState)
+
+handleAction _ Save = pure unit
+
+parseContract' :: String -> Maybe Contract
+parseContract' = fromTerm <=< hush <<< parseContract
+
+runAjax ::
+  forall m a.
+  ExceptT AjaxError (HalogenM State Action ChildSlots Void m) a ->
+  HalogenM State Action ChildSlots Void m (WebData a)
+runAjax action = RemoteData.fromEither <$> runExceptT action
+
+editorResize :: forall state action msg m. HalogenM state action ChildSlots msg m Unit
+editorResize = void $ query _marloweEditorPageSlot unit (Monaco.Resize unit)
+
+editorSetValue :: forall state action msg m. String -> HalogenM state action ChildSlots msg m Unit
+editorSetValue contents = void $ query _marloweEditorPageSlot unit (Monaco.SetText contents unit)
+
+editorGetValue :: forall state action msg m. HalogenM state action ChildSlots msg m (Maybe String)
+editorGetValue = query _marloweEditorPageSlot unit (Monaco.GetText identity)
+
+-- FIXME: This receives markers and sets errors and warnings. Maybe rename to processErrorsAndWarnings
+editorSetMarkers :: forall m. MonadEffect m => Array IMarker -> HalogenM State Action ChildSlots Void m Unit
+editorSetMarkers markers = do
+  let
+    warnings = filter (\{ severity } -> isWarning severity) markers
+
+    trimHoles =
```

Comment:
why do we do this again? maybe need a comment since I have forgotten!

Reply:
I just tried with `trimHoles` and without it. The difference is that without it the message is very lengthy. I'll add a comment , and I'll rename to `warningsWithTrimmedHoleMessage` to make it more explicit.

With `trimHoles`
<img width="1194" alt="Screen Shot 2020-12-30 at 13 23 54" src="https://user-images.githubusercontent.com/2634059/103367076-ee69ca80-4aa2-11eb-8c36-7af3f98eed4b.png">

Without
<img width="1200" alt="Screen Shot 2020-12-30 at 13 25 26" src="https://user-images.githubusercontent.com/2634059/103367089-f6296f00-4aa2-11eb-93ac-8f73fcfec132.png">